### ISSUE TITLE
Euro/American Truck Simulator 2 not working

### ISSUE THREAD

**Issue by hgouveia** [fholger/openvr_fsr#23]
Both Euro and American truck simulator does not work, 

VR mode starts and work fine, but FSR is not doing anything, this is the log i got for both

```
Initializing hooks...
Requested interface FnTable:IVRSystem_021
Requested interface FnTable:IVRCompositor_026
Shutting down hooks.
```

**Comment by mbgroot** [comment]
Acknowledged.
[openvr_mod.log](https://github.com/fholger/openvr_fsr/files/6836514/openvr_mod.log)

**Comment by fholger** [comment]
Could you give this custom build a quick spin, see if it does anything? I'd be interested in updated logs with this build.

[openvr_fsr_ts_test.zip](https://github.com/fholger/openvr_fsr/files/6872228/openvr_fsr_ts_test.zip)

**Comment by hgouveia** [comment]
> Could you give this custom build a quick spin, see if it does anything? I'd be interested in updated logs with this build.
> 
> [openvr_fsr_ts_test.zip](https://github.com/fholger/openvr_fsr/files/6872228/openvr_fsr_ts_test.zip)

Hello, just tried, but both games don't run/open anymore, but i got some logging

American Truck Simulator
```
Initializing hooks...
Requested interface FnTable:IVRSystem_021
Requested interface IVRSystem_021
Injecting GetRecommendedRenderTargetSize into IVRSystem_021
Requested interface FnTable:IVRCompositor_026
Requested interface IVRCompositor_026
Injecting Submit into IVRCompositor_026
```

Eurotruck simulator 2
```
Initializing hooks...
Requested interface FnTable:IVRSystem_021
Requested interface IVRSystem_021
Injecting GetRecommendedRenderTargetSize into IVRSystem_021
Requested interface FnTable:IVRCompositor_026
Requested interface IVRCompositor_026
Injecting Submit into IVRCompositor_026
```

**Comment by fholger** [comment]
Shame. Any particular error message? Do the games perhaps have their own log files that might contain further hints? Otherwise, I'm not really sure what else to do.

**Comment by hgouveia** [comment]
> Shame. Any particular error message? Do the games perhaps have their own log files that might contain further hints? Otherwise, I'm not really sure what else to do.

Hello, for the game itself nothing happened, just not open, not window, not console, nothing, but i was able to find some logging from the game itself

[ets2_game.crash.txt](https://github.com/fholger/openvr_fsr/files/6872368/ets2_game.crash.txt)
[ats_game.crash.txt](https://github.com/fholger/openvr_fsr/files/6872369/ats_game.crash.txt)

**Comment by fholger** [comment]
I may have simply done things in the wrong order. Please try this build.

[openvr_fsr_ts_test2.zip](https://github.com/fholger/openvr_fsr/files/6873679/openvr_fsr_ts_test2.zip)

**Comment by hgouveia** [comment]
> I may have simply done things in the wrong order. Please try this build.
> 
> [openvr_fsr_ts_test2.zip](https://github.com/fholger/openvr_fsr/files/6873679/openvr_fsr_ts_test2.zip)

Unfortunately, same issue the game does not run, 

in case it worth, the original build of openvr_fsr, doesnt crash the game, it manage to run, but it just not apply the FSR

[game.crash.txt](https://github.com/fholger/openvr_fsr/files/6873718/game.crash.txt)
[openvr_mod.log](https://github.com/fholger/openvr_fsr/files/6873720/openvr_mod.log)

**Comment by fholger** [comment]
Then I'm afraid this is as far as I can go on this issue. Unless someone who actually owns the games has the ability to debug the crash, there's nothing else I can do.

**Comment by hgouveia** [comment]
> Then I'm afraid this is as far as I can go on this issue. Unless someone who actually owns the games has the ability to debug the crash, there's nothing else I can do.

Yeap, i might be able to help, i am programmer, but i am not very proficient with openvr or c++, could you point me out, what i lines you think should i debug?

**Comment by fholger** [comment]
Familiarity with OpenVR is not required, but C++ would be helpful. You'd need to be able to build a debug version of the project, then start the game with that debug version through the debugger and let it tell you where the crash occurs. From the game's crash log I'm assuming that the crash is in the openvr project, and I'd need to know where it crashes and why. That's what the debugger would hopefully be able to tell us.

**Comment by hgouveia** [comment]
Hello @fholger, i managed to debug and found the issue, i hope this could be helpful

![image](https://user-images.githubusercontent.com/668866/126959211-aa2279f9-37a8-4739-a16d-a30bf1037894.png)

**Comment by fholger** [comment]
It is, thanks. I'll attach a fixed build later, let's see if that gets us any further :)

**Comment by hgouveia** [comment]
@fholger  FYI, after doing this, the game started and FSR was working , just as reference since i dont know is the best solution for this, because i feel i got lower native resolution, but attaching logs and screen in case is relevant to you


![image](https://user-images.githubusercontent.com/668866/126963240-d84fd77f-c3ae-46f5-b4a4-51dd82aa307f.png)
[openvr_mod.log](https://github.com/fholger/openvr_fsr/files/6876938/openvr_mod.log)

**Comment by fholger** [comment]
Yes, the log looks good, looks like it's working. Your fix is more or less what I'm going to do, just going to move the null checks around a bit :)

**Comment by hgouveia** [comment]
@fholger i've found another issue, when i open the game with openvr_fsr, it is set as **Standing Position**, and this game normally use **Seated Position** , when using standard **openvr_api.dll** it does starts as **Seated Position**, i was wondering if probably you could add a new entry to the **openvr_mod.cfg** where the users could specify the Tracking Universe, also found this in case is useful https://github.com/ValveSoftware/openvr/blob/master/headers/openvr.h#L4530-L4531 , or probaly just bypass that and not been overrided to  **Standing Position**

**Comment by luopeng69131** [comment]
@hgouveia Hi I'd like to know if you are willing to share the build.  Thank you!

**Comment by hgouveia** [comment]
> @hgouveia Hi I'd like to know if you are willing to share the build. Thank you!

@luopeng69131 lets wait until @fholger has a fix for this, since the game is running on **Standing Position** , that makes it imposible to play, because the view is wayyy outside the truck, very up , and reseting the view ingame with F12 or inside steamvr, doesnt fix it

**Comment by luopeng69131** [comment]
@hgouveia  ok but thank you anyway.

**Comment by fholger** [comment]
> 
> 
> @fholger i've found another issue, when i open the game with openvr_fsr, it is set as **Standing Position**, and this game normally use **Seated Position** , when using standard **openvr_api.dll** it does starts as **Seated Position**, i was wondering if probably you could add a new entry to the **openvr_mod.cfg** where the users could specify the Tracking Universe, also found this in case is useful https://github.com/ValveSoftware/openvr/blob/master/headers/openvr.h#L4530-L4531 , or probaly just bypass that and not been overrided to **Standing Position**

I'm afraid I'm at a loss here. The mod does not override the tracking universe, it doesn't even intercept that call. This is entirely in the hands of the game and should not depend on the used DLL, at all.

**Comment by hgouveia** [comment]
> > @fholger i've found another issue, when i open the game with openvr_fsr, it is set as **Standing Position**, and this game normally use **Seated Position** , when using standard **openvr_api.dll** it does starts as **Seated Position**, i was wondering if probably you could add a new entry to the **openvr_mod.cfg** where the users could specify the Tracking Universe, also found this in case is useful https://github.com/ValveSoftware/openvr/blob/master/headers/openvr.h#L4530-L4531 , or probaly just bypass that and not been overrided to **Standing Position**
> 
> I'm afraid I'm at a loss here. The mod does not override the tracking universe, it doesn't even intercept that call. This is entirely in the hands of the game and should not depend on the used DLL, at all.

yeap i also was looking a bit the code and i didnt found anything that set to standing position, so is very weird this happens, but it does, and when i use the original **openvr_api.dll** it starts straight into **Seated**

i try to force it adding this like, just to see if possible to modify

![devenv_zjxGl6AMRH](https://user-images.githubusercontent.com/668866/127019765-c74aa620-9261-4e65-9a24-471efacf3669.png)

but didnt work, so no clue at all


https://user-images.githubusercontent.com/668866/127020196-39cce914-870e-47db-80b9-40c3ea13c903.mp4

**Comment by fholger** [comment]
Well, for what it's worth, here's a test build with the crash fix. Let's see if other people face the same issue.
[openvr_fsr_ts_test3.zip](https://github.com/fholger/openvr_fsr/files/6879939/openvr_fsr_ts_test3.zip)

**Comment by gavmcn** [comment]
When I launch the game nothing happens and returns to the SteamVR Home.
[openvr_mod.log](https://github.com/fholger/openvr_fsr/files/6880539/openvr_mod.log)

**Comment by hgouveia** [comment]
> Well, for what it's worth, here's a test build with the crash fix. Let's see if other people face the same issue.
> [openvr_fsr_ts_test3.zip](https://github.com/fholger/openvr_fsr/files/6879939/openvr_fsr_ts_test3.zip)

[openvr_mod.log](https://github.com/fholger/openvr_fsr/files/6880606/openvr_mod.log)

Same as @gavmcn
> When I launch the game nothing happens and returns to the SteamVR Home.

I did some debugging, and the issue happens since `pnWidth` is nullptr but `pnHeight` is not, and make it crash, so i guess, the best solution is just apply the scaling calculation if both are not nullptr
![image](https://user-images.githubusercontent.com/668866/127050464-6654e89c-0338-4dfd-95a4-180945b97cdd.png)

**Comment by fholger** [comment]
It crashes in the `Log()` line, but that one's not active in the source. What happens if you comment it out, as it should be? The prior accesses are guarded, so that should not crash.

**Comment by hgouveia** [comment]
> It crashes in the `Log()` line, but that one's not active in the source. What happens if you comment it out, as it should be? The prior accesses are guarded, so that should not crash.

yeap you are right, i uncommented that line because before the only thing i got was this
![](https://user-images.githubusercontent.com/668866/127051666-a08aae8e-6f23-491d-8de4-54030dc42565.png)

but after uncommented that line, i got the info about pnWidth been nullptr, so i guess that those values are been use somewhere else that make it crash

also, i did the same fix , and this one did work, but to be honest dont know why if is almost the same
![image](https://user-images.githubusercontent.com/668866/127052220-705b2b4c-6f0d-48b0-a6ff-32086b378273.png)
[openvr_mod.log](https://github.com/fholger/openvr_fsr/files/6880923/openvr_mod.log)


BTW, the seated mode is working now

**Comment by fholger** [comment]
Sounds like the game is doing something really stupid with that call. I modified the fix for the v1.0 release. However, still don't think I can do anything about the tracking space.

**Comment by hgouveia** [comment]
@fholger just tested and working great, even the tracking space now, thank for this!

Closing this
